src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -150,8 +153,213 @@ CosineTree::CosineTree(const arma::mat& dataset,
   ConstructBasis(treeQueue);
 }
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+    dataset(new arma::mat(*other.dataset)),
+    delta(other.delta),
+    parent(NULL),
+    left(NULL),
+    right(NULL),
+    indices(other.indices),
+    l2NormsSquared(other.l2NormsSquared),
+    centroid(other.centroid),
+    basisVector(other.basisVector),
+    splitPointIndex(other.SplitPointIndex()),
+    numColumns(other.NumColumns()),
+    l2Error(other.L2Error()),
+    frobNormSquared(other.FrobNormSquared()),
+    localDataset(other.localDataset && other.parent == NULL)
```

Comment:
This isn't correct. `localDataset` indicates whether this `CosineTree` is responsible for deleting `dataset`. What you've written here does not match the initialization of `dataset` above.

Reply:
I'm not sure I follow. Are you saying that it should be initialized to `other.parent==NULL`? Put another way, will the copied tree be responsible for deleting its dataset even if the tree it copies from is not responsible for deleting its dataset? 
Sorry about the repeated questions I just want to make sure I fully understand what's going on. :)

Reply:
If the `CosineTree` has allocated the memory then it is responsible for deleting it. So, in this case, the copy constructor and copy operators are allocating new memory for the dataset and therefore that `CosineTree` object is responsible for deleting it.

Reply:
Ok, now we're getting closer with 7a8814f.  But now we need a bit of specialized behavior.  In a copy constructor, we have two situations: (a) the node is the root node, and (b) the node is not the root node.  If it's situation (a), then we need to copy the dataset and set `localDataset` to `true` so that the dataset is freed when the `CosineTree` is deallocated.  The code appears to do that now (I need to double-check but I'm pretty sure).

If it's situation (b) though, we need to *avoid* copying the dataset, as we are only holding a copy of the dataset in the root of the tree.  i.e., for a child node, `this->dataset == this->parent.dataset`, but the child node is not responsible for deleting the dataset, so `localDataset` should be `false`.  The code currently appears to set `localDataset` correctly, but `this->dataset` will still make a copy in this case.

In other trees, the way that this is done is that we manually set `left->dataset` and `right->dataset` after calling the copy constructor / constructor to create children.  Take a look at `BinarySpaceTree` and you'll see what I mean.

Thanks for persisting with this---I hope that my help has made sense and illuminated some details about pointers and allocations, and if not, don't hesitate to ask me to clarify. :+1:

Reply:
(a) Yes, I'm fairly sure that it ensures required behaviour there.

(b) I did think of the same issue but I ran this [sample](https://pastebin.com/YgtEX2Z8) and found that all the nodes of the tree point to the same memory location. I didn't quite understand why but seeing as it did what was needed I left it. 

So what am I doing wrong here?

Reply:
Oh I figured it out. I am in fact manually setting the dataset pointer just like in the `BinarySpaceTree`. However this assignment comes after allocating memory leading to a massive memory leak which I just found through `valgrind`. 

I'm curious, why didn't the Jenkins Memory Check job fail? Shouldn't it catch something like this?